doc/dev/transforms.rst

```diff
@@ -100,6 +100,119 @@ classical instructions as well as side-effecting operations (e.g., print) can al
+If this is your first time writing MLIR or LLVM passes, the boilerplate can be quite overwhelming. 
+Let's first set up the various boilerplate items required to register and run a new pass. 
+
+We'll create an empty pass in the ``Catalyst`` dialect that just prints out hello world to stdout. The full codelisting of the pass already exists in Catalyst as ``MyHelloWorldPass`` in the directories ``mlir/include/Catalyst/Transforms`` and ``mlir/lib/Catalyst/Transforms``, and we will go over it here. Note that the ``mlir/include`` (and ``mlir/lib``) directories consists of all the available dialects, so if you want to write a new pass in another dialect, it should be added to the subdirectory of that dialect. 
+
+The first thing to do is to create the pass object in the tablegen ``mlir/include/Catalyst/Transforms/Passes.td``:
+
+.. code-block::
+
+    def MyHelloWorldPass : Pass<"my-hello-world"> {
+        let summary = "An empty pass boilerplate that prints out hello world.";
+
+        let constructor = "catalyst::createMyHelloWorldPass()";
+    }
+
+When the dialect is built, this tablegen def will be built to a C++ file ``mlir/build/include/Catalyst/Transforms/Passes.h.inc``, containing the newly defined object called ``MyHelloWorldPassBase``, alongside the various necessary boilerplate methods in the MLIR infrastructure. Tablegen is designed such that we don't have to write all that boilerplate ourselves. 
+
+Now we write the pass itself. Create a new file ``mlir/lib/Catalyst/Transforms/MyHelloWorldPass.cpp`` with the following content:
+
+.. code-block:: cpp
+
+    #define DEBUG_TYPE "myhelloworld"
+
+    #include "Catalyst/IR/CatalystDialect.h"
+    #include "mlir/Pass/Pass.h"
+    #include "llvm/Support/Debug.h"
+
+    using namespace llvm;
+    using namespace mlir;
+    using namespace catalyst;
+
+    namespace catalyst {
+    #define GEN_PASS_DEF_MYHELLOWORLDPASS
+    #include "Catalyst/Transforms/Passes.h.inc"
+
+    struct MyHelloWorldPass : public impl::MyHelloWorldPassBase<MyHelloWorldPass> {
+        using impl::MyHelloWorldPassBase<MyHelloWorldPass>::MyHelloWorldPassBase;
+
+        void runOnOperation() override { llvm::errs() << "Hello world!\n"; }
+    };
+
+    std::unique_ptr<Pass> createMyHelloWorldPass() { return std::make_unique<MyHelloWorldPass>(); }
+
+    } // namespace catalyst
+
+We make the pass object ``MyHelloWorldPass``, which inherits from the base class ``MyHelloWorldPassBase`` that tablegen will build in the namespace ``impl``. The function that determines what your pass actually does is the ``void runOnOperation()``. Here all the pass does is print out ``"Hello world!\n"``. 
+
+(A sidenote on printing messages in MLIR: there are two major printing options in LLVM. The `more standard one <https://llvm.org/docs/ProgrammersManual.html#the-llvm-debug-macro-and-debug-option>`_ is ``dbgs()``, which only prints when a debug flag is set. The other option is the ``errs()`` used here, which will print no matter what.)
+
```

Comment:
Can you add a small paragraph about patterns? Even if not necessary for that example, maybe just mention that it exists and points to external docs.

Reply:
Considering the immediate next section is about pattern rewrites (the `MatchAndRewrite` and the `ApplyAndFoldGreedily`), I don't think this boilerplate section needs to mention it.